File: 2221-check-if-a-parentheses-string-can-be-valid/check-if-a-parentheses-string-can-be-valid.java
class Solution {
    public boolean canBeValid(String s, String locked) {
        int n = s.length();
        if (n % 2 == 1) return false; // Odd-length strings cannot be balanced

        // Check from left to right (treat '(' as open)
        if (!validate(s, locked, '(')) {
            return false;
        }

        // Check from right to left (treat ')' as open)
        if (!validate(new StringBuilder(s).reverse().toString(), 
                      new StringBuilder(locked).reverse().toString(), ')')) {
            return false;
        }

        return true;
    }

    private boolean validate(String s, String locked, char openChar) {
        int balance = 0;
        int unlocked = 0;

        for (int i = 0; i < s.length(); i++) {
            if (locked.charAt(i) == '0') {
                unlocked++; // Unlocked, can be used as either '(' or ')'
            } else {
                if (s.charAt(i) == openChar) {
                    balance++; // Increment balance for openChar
                } else {
                    balance--; // Decrement balance for closeChar
                }
            }

            // If balance + unlocked < 0, it means we have more closeChars than openChars
            // and we cannot fix it with unlocked characters
            if (balance + unlocked < 0) {
                return false;
            }
        }

        return true;
    }
}

File: 2221-check-if-a-parentheses-string-can-be-valid/check-if-a-parentheses-string-can-be-valid.py
class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        if len(s) % 2 == 1:
            return False
        r, a = 0, 0 
        for c, i in zip(s, locked):
            # print(c, i)
            if i == "0":
                a += 1
            else:
                if c == '(':
                    r += 1 
                else:
                    if r > 0:
                        r -= 1
                    elif a > 0:
                        a -= 1
                    else: return False
        
        s = s[::-1]
        locked = locked[::-1]
        print(s, locked)
        r, a = 0, 0 
        for c, i in zip(s, locked):
            if i == "0":
                a += 1
            else:
                if c == ')':
                    r += 1 
                else:
                    if r > 0:
                        r -= 1
                    elif a > 0:
                        a -= 1
                    else: return False
        
        return True